content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc

```diff
@@ -0,0 +1,424 @@
+    return re.getThread().getJavaName().startsWith("vert.x-eventloop") ||
+        re.getThread().getJavaName().startsWith("executor-thread");
+  }
+}
+----
+<1> Enable the `jdk.ObjectAllocationInNewTLAB` and `jdk.ObjectAllocationOutsideTLAB` JFR events
+<2> Every 10,000 events, wait for all the JFR events
+<3> Calculate the allocated memory, by summing up the TLAB allocations of all relevant threads by filtering out the TLAB events of the web application's relevant threads
+<4> Reset the event stream for the next iteration
+<5> Is this a TLAB event?
+<6> Get the new TLAB size in case of a newly allocated TLAB, otherwise the out of TLAB allocated object size
+<7> We're only interested in the web application's own threads, in particular ignoring the main thread which runs the HTTP client of the test
+
+Here are the numbers I got from running 100,000 invocations:
+
+[source]
+----
+Requests executed: 10000, memory allocated: 34096 bytes/request
+Requests executed: 20000, memory allocated: 31768 bytes/request
+Requests executed: 30000, memory allocated: 31473 bytes/request
+Requests executed: 40000, memory allocated: 31462 bytes/request
+Requests executed: 50000, memory allocated: 31547 bytes/request
+Requests executed: 60000, memory allocated: 31545 bytes/request
+Requests executed: 70000, memory allocated: 31537 bytes/request
+Requests executed: 80000, memory allocated: 31624 bytes/request
+Requests executed: 90000, memory allocated: 31703 bytes/request
+Requests executed: 100000, memory allocated: 31682 bytes/request
+----
+
+As we see, there's some warm-up phase during allocation rates still get down,
+but after ~20 K requests, the allocation per request is fairly stable,
+with a volatility of ~1% when averaged out over 10K requests.
+I.e. this initial phase should be excluded during the actual test.
+
+To emphasize the key part again, this allocation is per _request_, it is independent from wall clock time and thus is not dependent on the machine running the test (i.e. the test should behave the same when running on a developer laptop and on a CI machine), nor is it subject to volatility induced by other workloads running concurrently.
+
+Based on that, the actual test could look like so:
+
+[source, java]
+----
+@Test
+@EnableEvent("jdk.ObjectAllocationInNewTLAB")
+@EnableEvent("jdk.ObjectAllocationOutsideTLAB")
+public void retrieveTodo() throws Exception {
+  Random r = new Random();
+  HttpClient client = HttpClient.newBuilder().build();
+
+  for (int i = 1; i<= 20_000; i++) { // <1>
+    executeRequest(r, client);
+  }
+
+  jfrEvents.awaitEvents();
+  jfrEvents.reset();
+
+  for (int i = 1; i<= 10_000; i++) { // <2>
+    executeRequest(r, client);
+  }
+
+  jfrEvents.awaitEvents();
+
+  long sum = jfrEvents.filter(this::isObjectAllocationEvent)
+      .filter(this::isRelevantThread)
+      .mapToLong(this::getAllocationSize)
+      .sum();
+
+  assertThat(sum / 10_000).isLessThan(33_000); // <3>
```

Comment:
I'm wondering if it would make sense to define a lower threshold as well since in the text for <3> you talk about being *within a boundary*, so +/- ~1% ? Of course, you could argue that doing less allocations is "always" fine ...

Reply:
That's a great point actually, touching on something that was on my mind. Say allocation improves (new JDK, etc.) and then gets worse again, e.g. from 100 -> 60 -> 80. Right now, we wouldn't spot that regression from 60 to 80, due to that initial improvement. I'll add a note on that. Thanks!